[CBR-82] Server-side TLS client check

KtorZ opened the pull request:
## Description

Require and check client TLS certificates to talk to the wallet backend.
Note that the tip of this branch can be used to test [daedalus/#893](https://github.com/input-output-hk/daedalus/pull/893)

## Linked issue

[[CBR-82]](https://iohk.myjetbrains.com/youtrack/issue/CBR-82)


## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

parsonsmatt commented on stack.yaml:
```diff
@@ -132,6 +132,12 @@ packages:
+- location:
+    git: https://github.com/parsonsmatt/servant-quickcheck
+    commit: 38cd86bd2214dcf623b216f90d8fc9657549aed6
+  extra-dep: true
```
ah! we can use hackage version instead :) `servant-quickcheck-0.0.4.1` has this exposed: http://hackage.haskell.org/package/servant-quickcheck-0.0.4.1/docs/Servant-QuickCheck-Internal-QuickCheck.html#v:serverSatisfiesMgr

parsonsmatt commented on tools/src/gencerts/Configuration.hs:
```diff
@@ -0,0 +1,256 @@
+
+forEach :: [a] -> ((Int, a) -> b) -> [b]
+forEach xs fn =
+    zipWith (curry fn) [0..(length xs - 1)] xs
```
This can be `zipWith (curry fn) [0..] xs`. `zipWith` will truncate the result list to the shorter of the two lists you give it, so the infinite list is :ok_hand: , and in fact, you can do `take 10 (zipWith (+) [0..] [1..])` and the laziness works out fine -- if you try to take `length [0..]`, then it will never terminate!

KtorZ replied:
:ok_hand:

parsonsmatt commented on tools/src/gencerts/Main.hs:
```diff
@@ -0,0 +1,118 @@
+
+            <*> strOption (short 'c' <> long "configuration-file" <> metavar "FILEPATH"
+                <> help "Configuration file describing the PKI")
+    in do
```
this gets a bit of needless indentation -- you can have:

```haskell
main :: IO ()
main = do
    let 
        opts :: ParserInfo Command
        opts = ...

    Command{..} <- execParser opts

    tlsConfig <- ...
```

etc :)

parsonsmatt commented on tools/src/gencerts/README.md:
```diff
@@ -0,0 +1,320 @@
+# GenCerts
```
such docs :heart:

parsonsmatt commented on wallet-new/integration/CLI.hs:
```diff
@@ -9,28 +9,42 @@ module CLI
+    tlsClientCertPath <- strOption $
+        long        "tls-client-cert"
+        <> metavar  "FILEPATH"
+        <> help     "Path to TLS client public certificate"
```
:tada:

KtorZ commented:
:+1: Addressed comments above as well as:

- Fix a small typo (`--ca-cert` instead of `--cacert` in one of the doc)
- Use `--cacert`instead of `-k` in the demo scripts as discussed in a previous PR
- Remove Python build artifacts and add them to .gitignore to make sure they won't be committed again
